tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -1331,9 +1332,34 @@ public:
    template <typename... Columns>
    void Exec(unsigned int, Columns &... columns)
    {
-      fDisplayerHelper->AddRow(columns...);
-      if (!fDisplayerHelper->HasNext()) {
+      // The normal mode of operation: checking whether we still have entries
+      // to be displayed and adding the corresponding row to the RDisplay object
+      // accordingly. The RDisplay::AddRow method will decrease a counter of the
+      // RDisplay object, so that when the counter reaches 0 no more rows should
+      // be added from here. This counter is set at creation time with a default
+      // value of 5 and can be changed by the user when calling Display. This
+      // check should be done before the first call to AddRow, otherwise we
+      // could trigger an integer overflow of the fEntries counter inside the
+      // RDisplay object. This means that, in case the only operation of the
+      // graph is Display, we have to run one extra entry and check that there
+      // are no more entries to display, to then trigger the second condition
+      // and finally stop the graph. We pay that (small) cost to avoid the
+      // integer overflow.
+      if (fDisplayerHelper->HasNext()) {
+         fDisplayerHelper->AddRow(columns...);
+         // When no more entries should be displayed, we have to stop adding rows
+         // and at the same time signal that this node of the graph has finished
+         // its job. This is the only action helper who is also responsible for
+         // sending the signal through the StopProcessing method. We add a boolean
+         // flag internal to this class, to make sure StopProcessing is only called
+         // once. Without it, this method would be called at each entry that is
+         // processed after the first condition is false, leading to a buggy
+         // scenario where this helper could stop the entire computation graph
+         // by adding as many "stop" signals as the number of nodes of the graph.
+         // This would trigger the stop mechanism in RLoopManager.
+      } else if (!fStopCalled) {
          fPrevNode->StopProcessing();
+         fStopCalled = true;
```

Comment:
Thank you very much for the debugging and the fix. The fact that we need this amount of text to explain these lines suggests that we might want to simplify this further.

I think it would be better if here we could just write:

```cpp
template <typename... Columns>
void Exec(unsigned int, Columns &... columns)
{
   if (fEntriesToProcess == 0)
      return;

   fDisplayerHelper->AddRow(columns...);
   --fEntriesToProcess;
   if (fEntriesToProcess == 0) // we are done, send a one-time stop signal upstream
      fPrevNode->StopProcessing();
}
```

This way we can also avoid adding the extra boolean `fStopCalled` and we can completely remove the method `RDisplay::HasNext`.

The requirement for implementing things this way is to move `RDisplay::fEntries` to `RDisplayHelper::fEntriesToProcess`, which requires small changes to `RInterface::Display` and `BuildAction(..., ActionTags::Display)` to pass that parameter to `RDisplayHelper`'s ctor instead of `RDisplay`'s ctor.

Reply:
It may work, but I also need to add an overload of `CreateAction` that accepts also an extra `nEntries` parameter, but that seems too overkill to me

Reply:
Or we can just live with two instances of `fEntries`, one in the display and one in the helper. The one in the display is `const` and never touched, is just there to be passed to the constructor of the helper.

Reply:
> I also need to add an overload of CreateAction that accepts also an extra nEntries parameter

I don't think so, you should be able to pass extra things to `CreateAction` in the `helperArg` argument

Reply:
Fair enough, an `std::pair` it is then